### PR TITLE
CORE-502 Support for passphrase encryption of secrets

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1738,6 +1741,13 @@ ffi!(fn pgp_export_keydata(session: *mut Session,
 
     let mut keydata = Vec::new();
     if secret {
+        // No-op if there is no passphrase (NULL).
+        // If there is a passphrase, only encrypted parts are decrypted.
+        let cert = decrypt_cert(cert, passphrase)?;
+
+        // No-op if there is no passphrase (NULL).
+        let cert = encrypt_cert(cert, passphrase)?;
+
```
Comment (thgoebel-planck) [inline src/lib.rs]: This looks circular? This would make sense if the passphrases are different (like in `manage_passphrases`). But since the passphrases are the same, why is this needed?

File: src/manage_passphrase.rs [hunk 2]
```diff
@@ -69,6 +57,43 @@ ffi!(
...
+    if !have_passphrase {
+        // nothing to do
+        Ok(cert)
+    } else {
+        let passphrase = unsafe { check_cstr!(passphrase) }
+            .to_str()
+            .map_err(|_| illegal_value("new passphrase cannot be converted to string"))?;
+
+        let passphrase = Password::from(passphrase);
+        let encrypted_packets = encrypted_packets(&cert, &passphrase)?;
```
Comment (thgoebel-planck) [inline src/manage_passphrase.rs:71]:
```suggestion
        let encrypted_packets = encrypt_packets(&cert, &passphrase)?;
```

Same for `decrypted_packets`.

I find the naming confusing. Why is it `encrypt_cert()` but `encrypted_packets()`? Both functions do the same thing, just on different objects. So the naming should be aligned.

(I know it was like this before, but it would be an opportunity to fix it.)

File: src/lib.rs [hunk 3]
```diff
@@ -1738,6 +1741,13 @@ ffi!(fn pgp_export_keydata(session: *mut Session,
 
     let mut keydata = Vec::new();
     if secret {
+        // No-op if there is no passphrase (NULL).
+        // If there is a passphrase, only encrypted parts are decrypted.
```
Comment (thgoebel-planck) [inline src/lib.rs]:
```suggestion
```

What else, if not encrypted packets, would you decrypt? :) -> useless comment, remove

**Review by thgoebel-planck — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (ghost) [reply src/lib.rs]: I want to both support the case "not encrypted -> encrypted" and "encrypted -> encrypted (both with the same passphrase)" explicitly, and there is only one passphrase per account in our setup in this case.

File: src/lib.rs (hunk 3 above)
Comment (ghost) [reply src/lib.rs]: IMO this is only clear if you are an insider re certificates. Hence the comment. I want to stress that decrypting something that has not been encrypted in the first place is not an error.

File: src/manage_passphrase.rs (hunk 2 above)
Comment (ghost) [reply src/manage_passphrase.rs:71]: No, not on different objects, but result and behavior is different. From the naming alone, `encrypt_cert` indicates that a cert gets encrypted, and `encrypted_packets` yields the encrypted packets of something.

Comment (thgoebel-planck) [reply src/manage_passphrase.rs:71]: But `encrypt_cert` yields the encrypted cert, so let's call it `encrypted_cert`.
And `encrypted_packets` encrypts the packets, so let's call it `encrypt_packets`.

=> I still don't see the difference. Both functions encrypt X and then they yield the encrypted X. Both functions could reasonably be called `encrypt_X` and `encrypted_X`. So avoid confusion (I'm confused) we should use the same naming scheme.

File: src/lib.rs (hunk 1 above)
Comment (thgoebel-planck) [reply src/lib.rs]: Ah, make sense. Can you comment this at the top?

```
// (Re-)encrypt with the passphrase. This allows both cases to work:
// 1. not encrypted -> encrypted
// 2. encrypted -> encrypted (this is effectively a no-op)
```

File: src/lib.rs (hunk 3 above)
Comment (thgoebel-planck) [reply src/lib.rs]: Hm okay. Maybe this comment I [proposed above](https://github.com/plancksecurity/foundation-planckCoreSequoiaBackend/pull/15#discussion_r1716875579) is clearer and can replace this comment?

Alternatively, a clearer version would be:

> If there is a passphrase but the cert is unencrypted, this is also a no-op.

File: src/manage_passphrase.rs (hunk 2 above)
Comment (ghost) [reply src/manage_passphrase.rs:71]: It's natural language. `encrypt_cert` (verb, object) is a _transformation_. The original certificate may be consumed or not considered important anymore. `encrypted_packets` is a _getter_. In java I may have called it `getEncryptedPackets` and put it on a certificate. Wait, no, I don't think java allows such monkey patching. The object it operates on (certificate) is not touched by the operation. I can use `get_encrypted_packets_from_certificate` and the like if that helps you. Please also note that `encrypt_cert` is public, the other is purely internal at the moment.

File: src/lib.rs (hunk 3 above)
Comment (ghost) [reply src/lib.rs]: Will try to clarify this. I don't consider it obvious at all that you can decrypt something that was not encrypted in the first place. E.g. if you decrypt an unencrypted text file, you'll get garbage.

File: src/manage_passphrase.rs (hunk 2 above)
Comment (thgoebel-planck) [reply src/manage_passphrase.rs:71]: Ah, I missed that one is `&Cert` and the other one is `Cert`. Now it makes more sense.

File: src/lib.rs (hunk 3 above)
Comment (thgoebel-planck) [reply src/lib.rs]: > E.g. if you decrypt an unencrypted text file, you'll get garbage.

True, and I agree it's worth to document it. Just trying to find a better phrasing, since I needed your explanation here to understand the comment.

**Review by thgoebel-planck — APPROVED** [review]
Looks good, thanks for the clarifications.

**Review by saschabacardit — APPROVED** [review]
